proto/canto/csr/v1/tx.proto

```diff
@@ -2,12 +2,97 @@ syntax = "proto3";
+// Wrapper so that we can create a map to an array of Nonces in MsgRegisterCSR and MsgUpdateCSR
+message NoncesArray {
+    repeated uint64 nonces = 1;
+}
+
+// Defining a type for the canto addresses used across messages
+message CantoAddress {
+    string address = 1 [
+        (gogoproto.nullable) = false
+        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"
+    ];
+}
+
+// Defining an NFT array type so that we can nest it in a map
+message NFTArray {
+    repeated string nfts = 1;
+}
+
+
+// This message allows users to register a set of smart contracts into the CSR store. 
+// It will input the deployer address, number of NFTs to mint, NFT allocation per account 
+// if desired, set of smart contracts to start accepting transaction revenue, and nonces for address derivation.
+message MsgRegisterCSR {
+    // The deployer is the same account that signed the MsgRegisterCSR transaction
+    string deployer = 1;
+    // nft_supply tracks the number of NFTs the deployer wants to mint
+    uint64 nft_supply = 2;
+    // allocations keeps track of the allocations of NFTs that the deployer desires upon registration
+    map<string, uint64> allocations = 3;
+    // contracts are the smart contract addresses that we want to register with this CSR pool
+    repeated string contracts = 4;
+    // nonces is used for address derivation and deployment verification.
+    map<string, NoncesArray> nonces = 5;
+}
+
+// MsgRegisterCSRResponse returns no fields
+message MsgRegisterCSRResponse {
+}
+
+// This message allows users to update an existing CSR pool with additional smart contracts. 
+// It will input the deployer address, the existing CSR pool address, a list of smart contracts 
+// to add, and nonces for derivation.
+message MsgUpdateCSR {
+    // The deployer is the same account that signed the MsgUpdateCSR transaction
+    string deployer = 1;
+    // pool_address corresponds to the Canto address of the CSR pool 
+    // that is accumulating rewards for a given set of NFTs
+    CantoAddress pool_address = 2;
+    // contracts are EVM addresses of the new contracts to be added to this CSR
+    repeated string contracts = 3;
+    // nonces is used for address derivation and deployment verification
+    map<string, NoncesArray> nonces = 4;
+}
+
+
+// MsgUpdateCSRResponse returns no fields
+message MsgUpdateCSRResponse {
+}
+
+// Whenever a user, owning a set number of NFTs for a set of CSRs, would like to 
+// withdraw the rewards associated with each of their NFTs, they submit a MsgWithdrawCSR. 
+// The user submits the message with a set of CSRs that they would like to withdraw from, 
+// and the set of NFTs within each CSR that they will be withdrawing from.
+message MsgWithdrawCSR {
+    // The addresses of the csr_pools that the rewards will be withdrawn from
+    repeated CantoAddress csr_pools = 1;
+    // The identifiers of each NFT for each CSRPool
+    map<CantoAddress, NFTArray> pool_to_nfts = 2;
```

Comment:
CantoAddress does not need to be the key in the map, it can just be an array of array of NFTs

Reply:
I did map because itll be easier to index instead of using ints can just use the pool address as key